### PR TITLE
[Plinth] [Builtins] Fix 'writeBits'

### PR DESCRIPTION
**Pull request by effectfully**
This will hopefully eventually fix the issue of `writeBits` being unusable in Plinth.

### CODE REVIEW

**Comment by kwxm** [comment]
See #6331.

File: plutus-tx-plugin/test/Plugin/Profiling/9.6/argMismatch2.eval.golden
Comment (effectfully) [inline plutus-tx-plugin/test/Plugin/Profiling/9.6/argMismatch2.eval.golden:1]: Looks like maybe for `Profiling` tests we should turn off showing uniques.

**Comment by effectfully** [comment]
I'm labeling this with "No Changelog Required" since I don't think our users are interested in `writeBits` being fixed, given that it's not quite released properly yet.

**Review by kwxm — APPROVED** [review]
Yep, this seems to work.

File: plutus-tx/src/PlutusTx/Builtins/HasOpaque.hs [hunk 1]
```diff
@@ -204,6 +204,18 @@ instance HasFromOpaque BuiltinBool Bool where
     fromOpaque b = ifThenElse b True False
     {-# INLINABLE fromOpaque #-}
 
+instance HasToOpaque [BuiltinInteger] (BuiltinList BuiltinInteger) where
```
Comment (zliu41) [inline plutus-tx/src/PlutusTx/Builtins/HasOpaque.hs:207]: Why not `instance HasToOpaque a o => HasToOpaque [a] (BuiltinList o)`?

File: plutus-tx-plugin/test/Plugin/Debug/9.6/letFun.pir.golden [hunk 2]
```diff
@@ -8,19 +8,19 @@
       { no-src-span }
       (datatype
         { no-src-span }
-        (tyvardecl { no-src-span } Bool-445 ({ no-src-span } type))
+        (tyvardecl { no-src-span } Bool-451 ({ no-src-span } type))
```
Comment (zliu41) [inline plutus-tx-plugin/test/Plugin/Debug/9.6/letFun.pir.golden:11]: I thought we didn't end up printing uniques for golden files, but apparently we did.

File: plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs [hunk 3]
```diff
@@ -332,6 +334,12 @@ defineBuiltinTerms = do
     -- Text constant
     defineBuiltinTerm annMayInline 'Builtins.emptyString $ PIR.mkConstant annMayInline ("" :: Text)
 
+    -- List constants
+    defineBuiltinTerm annMayInline 'Builtins.mkNilInteger $
```
Comment (zliu41) [inline plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs:338]: `mkNilInteger` and `mkNilBool` are not builtins, so `defineBuiltinTerm` is both a misnomer and unnecessary. How about `` mkNil :: DefaultUni `Contains` a => Proxy a -> BuiltinList a ``, and handle it in the plugin similar to `&&` and `||`?

File: plutus-tx/src/PlutusTx/Builtins/HasOpaque.hs (hunk 1 above)
Comment (effectfully) [reply plutus-tx/src/PlutusTx/Builtins/HasOpaque.hs:207]: I wanted to make it that, but it's pretty hard! We need to have the fully specialized `a` and compute the right type tag from it and it seems like it's above my Template Haskell skills currently. I can probably make it work without TH, but it won't be reliable. In any case, I do intend to create a follow-up polishing this mess a bit, this PR was just for fixing the immediate issue.

File: plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs (hunk 3 above)
Comment (effectfully) [reply plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs:338]: I'm following the existing convention to a T, you can see that it's the same `defineBuiltinTerm` for the empty bytestring right above, I've only replicated that for the empty lists. 

> How about mkNil :: DefaultUni `Contains` a => Proxy a -> BuiltinList a, and handle it in the plugin similar to && and ||?

(see my other comment)

**Pull request merged**